reactor-core/src/test/java/reactor/core/publisher/AbstractOnDiscardShouldNotLeakTest.java

```diff
@@ -0,0 +1,443 @@
+            TestPublisher<Tracked<?>>[] testPublishers = new TestPublisher[subscriptionsNumber];
+
+            for (int i1 = 0; i1 < subscriptionsNumber; i1++) {
+                testPublishers[i1] = TestPublisher.createNoncompliant(TestPublisher.Violation.DEFER_CANCELLATION, TestPublisher.Violation.REQUEST_OVERFLOW);
+            }
+
+            Publisher<Tracked<?>> source = transform(testPublishers[0], Arrays.copyOfRange(testPublishers, 1, testPublishers.length));
+
+            if (conditional) {
+                if (source instanceof Flux) {
+                    source = ((Flux<Tracked<?>>)source).filter(t -> true);
+                } else {
+                    source = ((Mono<Tracked<?>>)source).filter(t -> true);
+                }
+            }
+
+            AssertSubscriber<Tracked<?>> assertSubscriber = new AssertSubscriber<>(Operators.enableOnDiscard(null, Tracked::safeRelease), 0);
+            if (fused) {
+                assertSubscriber.requestedFusionMode(Fuseable.ANY);
+            }
+            source.subscribe(assertSubscriber);
+
+            if (subscriptionsNumber == 1) {
+                Tracked<Integer> value = new Tracked<>(1);
+                RaceTestUtils.race(() -> RaceTestUtils.race(assertSubscriber::cancel, () -> assertSubscriber.request(Long.MAX_VALUE), scheduler), () -> testPublishers[0].next(value), scheduler);
+            } else {
+                int startIndex = --index[0];
+                Tracked<Integer> value1 = new Tracked<>(startIndex);
+                int secondIndex = --index[0];
+                Tracked<Integer> value2 = new Tracked<>(secondIndex);
+                Runnable action = () -> RaceTestUtils.race(() -> testPublishers[startIndex].next(value1), () -> testPublishers[secondIndex].next(value2), scheduler);
+
+                while (index[0] > 0) {
+                    int nextIndex = --index[0];
+                    Tracked<Integer> nextValue = new Tracked<>(nextIndex);
+                    Runnable nextAction = action;
+                    action = () -> RaceTestUtils.race(nextAction, () -> testPublishers[nextIndex].next(nextValue), scheduler);
+                }
+                RaceTestUtils.race(() -> RaceTestUtils.race(assertSubscriber::cancel, () -> assertSubscriber.request(Long.MAX_VALUE), scheduler), action, scheduler);
+            }
+
+            List<Tracked<?>> values = assertSubscriber.values();
+            values.forEach(Tracked::release);
+
+            Tracked.assertNoLeaks();
+        }
+        scheduler.dispose();
+    }
+
+    @Test
+    public void ensureNoLeaksPopulatedQueueAndRacingCancelAndOnNext() {
+        Assumptions.assumeThat(subscriptionsNumber())
+                .isOne();
+        Hooks.onNextDropped(Tracked::safeRelease);
+        Hooks.onErrorDropped(e -> {});
+        Hooks.onOperatorError((e,v) -> null);
+        Scheduler scheduler = Schedulers.newParallel("testScheduler", subscriptionsNumber() + 1);
+        scheduler.start();
```

Comment:
this block seems to be repeated in tests, perhaps it can be reused

Reply:
Moved to a common field which setups `@Before` each test and dispose in the `@After` method